lib/rspec/que/queue_up.rb

```diff
@@ -1,89 +1,155 @@
 require 'rspec/mocks/argument_list_matcher'
+require 'time'
 
 module RSpec
   module Que
     module Matchers
       class QueueUp
         def initialize(job_class = nil)
+          @matchers = [QueuedSomething.new]
+          @matchers << QueuedClass.new(job_class) if job_class
           @job_class = job_class
+
+          @stages = nil
         end
 
         def matches?(block)
           @before_jobs = enqueued_jobs.dup
           block.call
-          enqueued_something? && enqueued_correct_class? && with_correct_args?
+
+          @stages = []
+          @matched_jobs = new_jobs
+          @matchers.each do |matcher|
+            @stages << {matcher: matcher, candidates: @matched_jobs.dup}
+            @matched_jobs = @matched_jobs.delete_if {|job| !matcher.matches?(job) }
+          end
+          !@matched_jobs.empty?
         end
 
         def with(*args)
-          raise "Must specify the job class when specifying arguments" unless job_class
+          @matchers << QueuedArgs.new(args)
+          self
+        end
 
-          @argument_list_matcher = RSpec::Mocks::ArgumentListMatcher.new(*args)
+        def at(the_time)
+          matcher = QueuedAt.new(the_time)
+          @matchers << matcher
           self
         end
 
         def failure_message
-          unless enqueued_something?
-            return "expected to enqueue a #{job_class || 'job'}, enqueued nothing"
-          end
-
-          unless enqueued_correct_class?
-            return "expected to queue up a #{job_class}, " \
-                   "enqueued a #{enqueued_jobs.last[:job_class]}"
-          end
+          # last stage to have any candidate jobs
+          failed_stage = @stages.select {|s| !s[:candidates].empty? }.last || @stages.first
```

Comment:
So basically just in the case:
`.to queue_up`? Otherwise there should be another stage detected as failed.
I feel this should then rather be detected by the `QueuedSomething` matcher.